arrow/obsidian.c

```
@@ -158,18 +170,47 @@ evolve(MTState *random_state,
-      for (reactant = 0; reactant < reactants_lengths[reaction]; reactant++) {
-        index = reactants_indexes[reaction] + reactant;
-        count = outcome[reactants[index]];
-        propensities[reaction] *= choose(count, reactions[index]);
+
+      // Compute propensity depending on the reaction's form
+      switch(forms[reaction]) {
+
+        case 0: // Standard Gillespie
+          // Initialize the propensity for reaction with its rate constant
+          propensities[reaction] = rates_flat[reaction];
+
+          // Go through each reactant and calculate its contribution to the
+          // propensity based on the counts of the corresponding substrate,
+          // which is then multiplied by the reaction's original rate and the
+          // contributions from other reactants
+          for (reactant = 0; reactant < reactants_lengths[reaction]; reactant++) {
+            index = reactants_indexes[reaction] + reactant;
+            count = outcome[reactants[index]];
+            propensities[reaction] *= choose(count, reactions[index]);
+          }
+
+          break;
+
+        case 1: // Michaelis-Menten
+          // Initialize the propensity for reaction with its maximal propensity
+          index = rates_indexes[reaction];
+          propensities[reaction] = rates_flat[index];
+
+          // Go through each reactant and calculate its contribution to the
+          // propensity based on the counts of the corresponding substrates
+          // relative to their Michaelis-Menten constant.
+          for (reactant = 0; reactant < reactants_lengths[reaction]; reactant++) {
+            index = reactants_indexes[reaction] + reactant;
+            count = outcome[reactants[index]];
+            propensities[reaction] *= fraction_saturation(count,
+              rates_flat[rates_indexes[reaction] + reactant + 1]);
+          }
+
+          break;
+
+        default:
+          printf("arrow.obsidian.evolve - unexpected form: %ld", forms[reaction]);
```

This probably needs to throw an error rather than just print a message.